### PR TITLE
Exempt `kube-system` namespace in `PodSecurity` admission config

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:
-  For clusters v1.23+ (where the new `PodSecurity` plugin was promoted to beta), add `PodSecurity` to default plugins
- If user has set a config in the shoot spec for this plugin, `kube-system` is added to exempted namespaces and every other config is untouched.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/5250

**Special notes for your reviewer**:

**Release note**:

```other operator
If a config for `PodSecurity` admission plugin is provided in the Shoot spec, `kube-system` is added to the exempted namespace.
```

### CODE REVIEW

File: pkg/operation/botanist/kubeapiserver.go [hunk 1]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+func (b *Botanist) ensureAdmissionPluginConfig(plugins []gardencorev1beta1.AdmissionPlugin, kubernetesVersion string) ([]gardencorev1beta1.AdmissionPlugin, error) {
+	var (
+		index                     int
+		allowPrivilegedContainers = pointer.BoolDeref(b.Shoot.GetInfo().Spec.Kubernetes.AllowPrivilegedContainers, false)
+		pspDisabled               = b.Shoot.PSPDisabled
+	)
+
+	for i, plugin := range plugins {
+		if plugin.Name == "PodSecurity" {
+			index = i
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go:196]: You can probably `break` here?

File: pkg/operation/botanist/kubeapiserver.go [hunk 2]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+		}
+	}
+
+	// if a config is set in the shoot spec
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
	// If a config for the PodSecurity plugin is set in the shoot spec, then ...? <- do what?
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 3]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+	if plugins[index].Config != nil {
+		config, err := runtime.Decode(codec, plugins[index].Config.Raw)
+		if err != nil {
+			return plugins, err
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: Typically we return `nil` when an error occurs

```suggestion
			return nil, err
```

(same below)

File: pkg/operation/botanist/kubeapiserver.go [hunk 4]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+		}
+		admConfig, ok := config.(*admissionapiv1beta1.PodSecurityConfiguration)
+		if !ok {
+			return plugins, fmt.Errorf("PodSecurity admission config is not of type PodSecurityConfiguration")
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
			return plugins, fmt.Errorf("expected admissionapiv1beta1.PodSecurityConfiguration but got %T", config)
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 5]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+		}
+
+		// exempt kube-system namespace
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: Please adapt the PR description accordingly.

File: pkg/operation/botanist/kubeapiserver.go [hunk 6]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+		if !slices.Contains(admConfig.Exemptions.Namespaces, metav1.NamespaceSystem) {
+			admConfig.Exemptions.Namespaces = append(admConfig.Exemptions.Namespaces, metav1.NamespaceSystem)
+		}
+		// if allowPrivilegedContainers is false, enforce restricted level
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
		// if allowPrivilegedContainers is false, enforce restricted level by default
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 7]
```diff
@@ -156,6 +191,72 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+				},
+				Exemptions: admissionapiv1beta1.PodSecurityExemptions{
+					Namespaces: []string{metav1.NamespaceSystem},
+				},
+			}
+			defaultConfigData, err := runtime.Encode(codec, defaultConfig)
+			if err != nil {
+				return plugins, err
+			}
+			plugins[index].Config = &runtime.RawExtension{Raw: defaultConfigData}
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: This is a lot of duplication with the above branch. Can you harmonize this and DRY?

File: go.mod [hunk 8]
```diff
@@ -59,6 +59,7 @@ require (
 	k8s.io/kube-proxy v0.24.3
 	k8s.io/kubelet v0.24.3
 	k8s.io/metrics v0.24.3
+	k8s.io/pod-security-admission v0.24.3
```
Comment (rfranzke) [inline go.mod]: Should be 0.24.4 now with `master` branch

File: docs/usage/pod-security.md [hunk 9]
```diff
@@ -8,8 +8,44 @@ admissionPlugins:
...
-in `kubernetes.kubeAPIServer.AdmissionPlugins` field in the shoot spec. Please refer the example shoot manifest [here](../../example/90-shoot.yaml).
+in `kubernetes.kubeAPIServer.admissionPlugins` field in the shoot spec. Please refer the example shoot manifest [here](../../example/90-shoot.yaml).
 
 Only if this field is set, the cluster can be upgraded to kubernetes `v1.25`.
 
 <strong>Note:</strong> You should disable the admission plugin and wait till Gardener finish at least one shoot reconciliation before upgrading to `v1.25`. This is to make sure all the `PodSecurityPolicy` related resources deployed by Gardener are cleaned up.
+
+## Admission configuration for the `PodSecurity` admission plugin
+
+If `.spec.kubernetes.allowPrivilegedContainers` is set to false in the Shoot spec, and the `PodSecurityPolicy` admission plugin is disabled, an admission configuration for the `PodSecurity` admission plugin with `restricted` level as default is applied. If you wish to add your own custom configuration, you can do so in the Shoot spec under `.spec.kubernetes.kubeAPIServer.admissionPlugins` by adding:
```
Comment (rfranzke) [inline docs/usage/pod-security.md]:
```suggestion
If `.spec.kubernetes.allowPrivilegedContainers` is set to `false` in the Shoot spec and the `PodSecurityPolicy` admission plugin is disabled, an admission configuration for the `PodSecurity` admission plugin with `restricted` level as default is applied. If you wish to add your own custom configuration, you can do so in the Shoot spec under `.spec.kubernetes.kubeAPIServer.admissionPlugins` by adding:
```

File: docs/usage/pod-security.md [hunk 10]
```diff
@@ -8,8 +8,44 @@ admissionPlugins:
...
+      warn-version: "latest"
+    exemptions:
+      # Array of authenticated usernames to exempt.
+      usernames: []
+      # Array of runtime class names to exempt.
+      runtimeClasses: []
+      # Array of namespaces to exempt.
+      namespaces: []
+```
+All gardener deployed workloads are exempted in any case.
```
Comment (rfranzke) [inline docs/usage/pod-security.md]: Update this: Everything in `kube-system` namespace.

File: docs/usage/pod-security.md (hunk 9 above)
Comment (rfranzke) [reply docs/usage/pod-security.md]: You could make more clear that end-users cannot overwrite the defaults when `allowPrivilegedContainers=false`.

**Comment by shafeeqes** [comment]
/hold
I've some testing left to do

File: go.mod [hunk 11]
```diff
@@ -66,6 +66,8 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require k8s.io/pod-security-admission v0.24.4
```
Comment (rfranzke) [inline go.mod]: Can you please add this to above `require ()` block?

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 12]
```diff
@@ -16,30 +16,101 @@ package shoot
...
-	f := defaultShootCreationFramework()
-	f.Shoot = e2e.DefaultShoot("unpriv-")
+	f := framework.NewShootCreationFramework(&framework.ShootCreationConfig{
+		GardenerConfig: &framework.GardenerConfig{
+			ProjectNamespace:   projectNamespace,
+			GardenerKubeconfig: os.Getenv("KUBECONFIG"),
+			SkipAccessingShoot: false,
+		},
+	})
+
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go]:
```suggestion
	f := defaultShootCreationFramework()
	f.GardenerConfig.SkipAccessingShoot = false
```

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 13]
```diff
@@ -16,30 +16,101 @@ package shoot
...
+	f.Shoot = e2e.DefaultShoot("")
+	f.Shoot.Name = "e2e-unprivileged"
 	f.Shoot.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(false)
+	f.Shoot.Spec.Kubernetes.KubeAPIServer = &gardencorev1beta1.KubeAPIServerConfig{
+		AdmissionPlugins: []gardencorev1beta1.AdmissionPlugin{
+			{
+				Name:   "PodSecurity",
+				Config: getConfig(),
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go]: This function seems to be useless, you can just inline the config here.

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 14]
```diff
@@ -16,30 +16,101 @@ package shoot
...
+	  "defaults": {
+	    "enforce": "restricted",
+	    "enforce-version": "latest"
+	  }
+    }`)
+
+	return &runtime.RawExtension{Raw: admissionConfigData}
+}
+
+func getPod(namespace string) *corev1.Pod {
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go]:
```suggestion
func newPodForNamespace(namespace string) *corev1.Pod {
```

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 15]
```diff
@@ -16,30 +16,101 @@ package shoot
...
 		defer cancel()
+
 		Expect(f.CreateShootAndWaitForCreation(ctx, false)).To(Succeed())
 		f.Verify()
 
+		shootClient := f.ShootFramework.ShootClient.Client()
+
+		By("Create pod in the kube-system namespace")
+		podInKubeSystem := getPod(metav1.NamespaceSystem)
+		Expect(shootClient.Create(ctx, podInKubeSystem)).To(Succeed())
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go]:
```suggestion
		Expect(shootClient.Create(ctx, newPodForNamespace(metav1.NamespaceSystem))).To(Succeed())
```

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 16]
```diff
@@ -16,30 +16,101 @@ package shoot
...
+
+		By("Create pod in the default namespace")
+		podInDefault := getPod(metav1.NamespaceDefault)
+		err := shootClient.Create(ctx, podInDefault)
+		Expect(err).To(BeForbiddenError())
+		Expect(err).To(MatchError(ContainSubstring("pods %q is forbidden: violates PodSecurity %q", podInDefault.Name, "restricted:latest")))
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go]:
```suggestion
		Expect(shootClient.Create(ctx, newPodForNamespace(metav1.NamespaceDefault))).To(And(
			BeForbiddenError(),
			MatchError(ContainSubstring("pods %q is forbidden: violates PodSecurity %q", podInDefault.Name, "restricted:latest")),
		))
```

File: test/e2e/shoot/create_and_delete_unprivileged.go [hunk 17]
```diff
@@ -16,30 +16,101 @@ package shoot
...
+					Image: "nginx:1.14.2",
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: 80,
+						},
+					},
+				},
+			},
+		},
+	}
```
Comment (rfranzke) [inline test/e2e/shoot/create_and_delete_unprivileged.go:102]: Why is this pod spec causing a forbidden error? What is the problematic part?

File: pkg/operation/botanist/kubeapiserver.go [hunk 18]
```diff
@@ -54,7 +86,8 @@ func (b *Botanist) DefaultKubeAPIServer(ctx context.Context) (kubeapiserver.Inte
 	}
 
 	var (
-		apiServerConfig = b.Shoot.GetInfo().Spec.Kubernetes.KubeAPIServer
+		apiServerConfig   = b.Shoot.GetInfo().Spec.Kubernetes.KubeAPIServer
+		kubernetesVersion = b.Shoot.GetInfo().Spec.Kubernetes.Version
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: You can use `b.Shoot.KubernetesVersion` which also is `*semver.Version`, then you don't need to pass it again and again.

File: pkg/operation/botanist/kubeapiserver.go [hunk 19]
```diff
@@ -153,6 +191,69 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+		pspDisabled &&
+		!allowPrivilegedContainers {
+		admissionConfig.Defaults = admissionapiv1beta1.PodSecurityDefaults{
+			Enforce:        "restricted",
+			EnforceVersion: "latest",
+		}
+	} else {
+		// if allowPrivilegedContainers is true and user also hasn't specified the config, don't set any config for PodSecurity
+		if !userHasSetConfig {
+			plugins[index].Config = nil
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: Why is this line needed? 👀

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 20]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+					shootCopy = botanist.Shoot.GetInfo().DeepCopy()
+					shootCopy.Spec.Kubernetes = gardencorev1beta1.Kubernetes{
+						KubeAPIServer: &gardencorev1beta1.KubeAPIServerConfig{
+							AdmissionPlugins: []gardencorev1beta1.AdmissionPlugin{},
+						},
+						Version: "1.23.0",
+					}
+					botanist.Shoot.PSPDisabled = true
+				})
+				JustBeforeEach(func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go:485]:
```suggestion

				JustBeforeEach(func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 21]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+						}
+					}
+				})
+
+				Context("When the config is nil", func() {
+					Context("allowPrivilegedContainers=false", func() {
+						BeforeEach(func() {
+							shootCopy.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(false)
+						})
+						It("should set the PodSecurity admission config", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion

						It("should set the PodSecurity admission config", func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 22]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+							}
+							Expect(config).To(Equal(defaultConfig))
+						})
+					})
+
+					Context("allowPrivilegedContainers=true", func() {
+						BeforeEach(func() {
+							shootCopy.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(true)
+						})
+						It("should not set the PodSecurity admission config", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion

						It("should not set the PodSecurity admission config", func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 23]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+				})
+
+				Context("When the config is not nil", func() {
+					var (
+						err       error
+						ok        bool
+						config    runtime.Object
+						admConfig *admissionapiv1beta1.PodSecurityConfiguration
+					)
+					BeforeEach(func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go:513]:
```suggestion

					BeforeEach(func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 24]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+exemptions:
+  usernames: ["admin"]
+  runtimeClasses: ["random"]
+  namespaces: ["random"]
+`),
+								},
+							},
+						}
+					})
+					JustBeforeEach(func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion

					JustBeforeEach(func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 25]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+
+						admConfig, ok = config.(*admissionapiv1beta1.PodSecurityConfiguration)
+						Expect(ok).To(BeTrue())
+					})
+
+					Context("allowPrivilegedContainers=false", func() {
+						BeforeEach(func() {
+							shootCopy.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(false)
+						})
+						It("should add kube-system to exempted namespaces and enforce restricted level by default", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion

						It("should add kube-system to exempted namespaces and enforce restricted level by default", func() {
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 26]
```diff
@@ -461,6 +461,146 @@ var _ = Describe("KubeAPIServer", func() {
...
+							Expect(admConfig.Defaults.Warn).To(Equal("privileged"))
+							Expect(admConfig.Exemptions.Namespaces).To(ContainElement("kube-system"))
+						})
+					})
+
+					Context("allowPrivilegedContainers=true", func() {
+						BeforeEach(func() {
+							shootCopy.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(true)
+						})
+						It("should add kube-system to exempted namespaces and not touch other fields", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]:
```suggestion

						It("should add kube-system to exempted namespaces and not touch other fields", func() {
```

File: test/e2e/shoot/create_and_delete_unprivileged.go (hunk 17 above)
Comment (shafeeqes) [reply test/e2e/shoot/create_and_delete_unprivileged.go:102]: When `restricted` profile is set, 
```
Error from server (Forbidden): error when creating "sample-pod.yaml": pods "nginx" is forbidden: violates PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
```

Comment (rfranzke) [reply test/e2e/shoot/create_and_delete_unprivileged.go:102]: Oh wow, this is indeed quite restrictive. I am asking myself whether we should default `allowPrivilegedContainers=false` for `1.25+` shoots (#6570). It rather feels as if this should be a conscious decision by the cluster owner?

Comment (shafeeqes) [reply test/e2e/shoot/create_and_delete_unprivileged.go:102]: We could also do `baseline` level, which gives the minimal privileges. (The above pod can pass)
But the motive behind this particular e2e test is, even if the user sets a config (`restricted` or whatever) for the new plugin, we should be able to deploy our pods.

File: pkg/operation/botanist/kubeapiserver.go [hunk 27]
```diff
@@ -153,6 +189,68 @@ func (b *Botanist) computeKubeAPIServerAdmissionPlugins(defaultPlugins, configur
...
+	if !slices.Contains(admissionConfig.Exemptions.Namespaces, metav1.NamespaceSystem) {
+		admissionConfig.Exemptions.Namespaces = append(admissionConfig.Exemptions.Namespaces, metav1.NamespaceSystem)
+	}
+
+	// if allowPrivilegedContainers is false, overwrite Defaults and enforce restricted level
+	if versionutils.ConstraintK8sGreaterEqual123.Check(b.Shoot.KubernetesVersion) &&
+		pspDisabled &&
+		!allowPrivilegedContainers {
+		admissionConfig.Defaults = admissionapiv1beta1.PodSecurityDefaults{
+			Enforce:        "restricted",
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: As discussed, let's rather use `baseline` as default here. Please adapt the docs.

**Comment by shafeeqes** [comment]
I have reworked the PR as discussed, PTAL! @rfranzke

File: docs/usage/pod-security.md [hunk 28]
```diff
@@ -13,3 +13,39 @@ in `spec.kubernetes.kubeAPIServer.admissionPlugins` field in the `Shoot` resourc
 Only if the `PodSecurityPolicy` admission plugin is disabled the cluster can be upgraded to `v1.25`.
 
 > :warning: You should disable the admission plugin and wait until Gardener finish at least one `Shoot` reconciliation before upgrading to `v1.25`. This is to make sure all the `PodSecurityPolicy` related resources deployed by Gardener are cleaned up.
+
+## Admission configuration for the `PodSecurity` admission plugin
+
+If you wish to add your custom configuration for the `PodSecurity` plugin, you can do so in the Shoot spec under `.spec.kubernetes.kubeAPIServer.admissionPlugins` by adding:
+```yaml
```
Comment (rfranzke) [inline docs/usage/pod-security.md:21]:
```suggestion

```yaml
```

File: docs/usage/pod-security.md [hunk 29]
```diff
@@ -13,3 +13,39 @@ in `spec.kubernetes.kubeAPIServer.admissionPlugins` field in the `Shoot` resourc
 Only if the `PodSecurityPolicy` admission plugin is disabled the cluster can be upgraded to `v1.25`.
 
 > :warning: You should disable the admission plugin and wait until Gardener finish at least one `Shoot` reconciliation before upgrading to `v1.25`. This is to make sure all the `PodSecurityPolicy` related resources deployed by Gardener are cleaned up.
+
+## Admission configuration for the `PodSecurity` admission plugin
```
Comment (rfranzke) [inline docs/usage/pod-security.md]:
```suggestion
## Admission Configuration For The `PodSecurity` Admission Plugin
```

File: docs/usage/pod-security.md [hunk 30]
```diff
@@ -13,3 +13,39 @@ in `spec.kubernetes.kubeAPIServer.admissionPlugins` field in the `Shoot` resourc
...
+      warn-version: "latest"
+    exemptions:
+      # Array of authenticated usernames to exempt.
+      usernames: []
+      # Array of runtime class names to exempt.
+      runtimeClasses: []
+      # Array of namespaces to exempt.
+      namespaces: []
+```
+For proper functioning of Gardener, if not present, `kube-system` namespace will be added exempted to the exempted namespaces by Gardener.
```
Comment (rfranzke) [inline docs/usage/pod-security.md]:
```suggestion

For proper functioning of Gardener, `kube-system` namespace will also be automatically added to the `exemptions.namespaces` list.
```

**Review by rfranzke — COMMENTED** [review]
Very nice @shafeeqes !

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve
